first draft sop (#32)
aerogear/mobile-developer-console-operator: SOP/SOP-operator.adoc

ifdef::env-github[]
:status:
:tip-caption: :bulb:

> :note-caption: :information_source:

Reviewer: Missing:

- Prometheus Alerts Procedures
- Validate
- Monitoring

See here: https://github.com/aerogear/unifiedpush-operator/blob/master/SOP/SOP-operator.adoc#prometheus-alerts-procedures

Author: About the Prometheus Alerts Procedures, it seems to me that that is duplicating the Overview section. It's pretty much the same sentence and set of links so perhaps we could include one rather than both.

I renamed the `Validate` section to `Check Operator` so it's still in there just under a different name, the link to it references `Check Operator` rather than validate so it does work. 

The Monitoring section in ups includes the following text
```
== Monitoring

If the https://github.com/aerogear/unifiedpush-operator#monitoring-service-metrics[Monitoring Service (Metrics)] is enabled for the installation, a Grafana Dashboard titled `UnifiedPush Operator`, and the Prometheus Monitoring instance are created.
```
so I will include similar text for that here.

Author: Actually I'm wondering now if we need to include this text at all. If the user has landed on the SOP then probably the monitoring stuff is already in place. wdyt about removing this?

Reviewer: Hi @laurafitzgerald,

Following the comments inline.  

> About the Prometheus Alerts Procedures, it seems to me that that is duplicating the Overview section. It's pretty much the same sentence and set of links so perhaps we could include one rather than both.

Hi @laurafitzgerald it is not. It is what makes clear what it means the critical and etc and was a request from SRE's @Nanyte25. Also, we should have a standard for that, so no make sense i just this one be different. 

> I renamed the Validate section to Check Operator so it's still in there just under a different name, the link to it references Check Operator rather than validate so it does work.

I prefer the nomenclature `Validate` which IMHO makes more clear what is the purpose of the section. Also, it came from an example provided by @Nanyte25 and I do not think that we should do things differently here. Keep the same standard is helpful for us and admins to know how to use it and have a better understanding. 

> Actually I'm wondering now if we need to include this text at all. If the user has landed on the SOP then probably the monitoring stuff is already in place. wdyt about removing this?

It allows who is checking it and has no idea over the project and how it works to get the understanding over it.

mobile-developer-console-operator-58c8877fd8-g6dvr          1/1       Running   3          9d
----

If the deployment is present but it's scaled down to 0 pods, attempt to scale up the operator pod by running `oc scale --replicas=1 deployment mobile-developer-console-operator`

== Validate

. Switch to the Mobile Developer Console namespace by running `oc project <namespace>`  E.g `oc project mobile-developer-console`.
. Check if the operator pod to is present by running `oc get pods | grep operator`.
+
[source,shell]
----
$ oc get pods | grep operator
NAME                                           READY     STATUS    RESTARTS   AGE
mdc-operator-58c8877fd8-g6dvr                  1/1       Running   3          9d
----

+
NOTE: The operator is responsible for managing and creating all objects required to have the Mobile Developer Console Service deployed on the cluster.
. Check its logs by running `oc logs <operator-podname>`
NOTE: You can save the logs by running `oc logs <operator-podname> > <filename>.log`. The logs may provide you with useful information to lead you to the root cause, and they are also useful for providing to the project maintainers when you create an issue.

> +

Reviewer: ```suggestion
== Monitoring

If the https://github.com/aerogear/mobile-developer-console-operator#monitoring-service-metrics[Monitoring Service (Metrics)] is enabled for the installation, a Grafana Dashboard titled `MDC Operator`, and the Prometheus Monitoring instances are created.
```

Author: 👍

ifdef::env-github[]
:status:
:tip-caption: :bulb:
:note-caption: :information_source:
:important-caption: :heavy_exclamation_mark:
:caution-caption: :fire:
:warning-caption: :warning:
:table-caption!:
endif::[]

:toc:
:toc-placement!:

= Mobile Developer Console Operator- Standard Operating Procedures

:toc:
toc::[]

== Overview

The following guide outlines the steps required to manage and solve issues in the https://github.com/aerogear/mobile-developer-console-operator[Mobile Developer Console Operator].

== Reference Articles

- https://github.com/aerogear/mobile-developer-console[Mobile Developer Console]
- https://github.com/aerogear/mobile-developer-console-operator[Mobile Developer Console Operator]
- https://prometheus.io/docs/practices/alerting/[Prometheus Alerts documentation]

== Success Indicators

All alerts should appear as green in the Prometheus Alert Monitoring.

== Critical 

=== MobileDeveloperConsoleOperatorDown

. Switch to the Mobile Developer Console namespace by running `oc project <namespace>`  E.g `oc project mobile-developer-console`.

> . Check if the operator pod to is present by running `oc get pods | grep operator`.

Reviewer 2: ```suggestion
. Check if the operator pod is present by running `oc get pods | grep operator`.
```

+
[source,shell]
----
$ oc get pods | grep operator
mobile-developer-console-operator-58c8877fd8-g6dvr          1/1       Running   3          9d
----

> If the deployment is present, attempt to scale up the operator pod by running `oc `

Reviewer 2: ```suggestion
If the deployment is present but it's scaled down to 0 pods, attempt to scale up the operator pod by running `oc scale --replicas=1 deployment mobile-developer-console-operator`
```

=== Warning

==== MobilDeveloperConsolePodCPUHigh

This alert indicates that the Service pod(s) is/are using more CPU than expected.

> . Please following the <<To capture the logs>> procedure in order to capture the required information to send it to its maintainers.

Reviewer 2: ```suggestion
. Please follow the <<To capture the logs>> procedure in order to capture the required information to send it to its maintainers.
```

> . Following the steps <<To scale the pod>> in order to try to solve performance issues.

Reviewer 2: ```suggestion
. Follow the steps <<To scale the pod>> in order to try to solve performance issues.
```

==== MobilDeveloperConsolePodMemoryHigh

This alert indicates that the Service pod(s) is/are using more Memory than expected.

> . Please following the <<To capture the logs>> procedure in order to capture the required information to send it to its maintainers.

Reviewer 2: ```suggestion
. Please follow the <<To capture the logs>> procedure in order to capture the required information to send it to its maintainers.
```

> . Following the steps <<To scale the pod>> in order to try to solve performance issues.

Reviewer 2: ```suggestion
. Follow the steps <<To scale the pod>> in order to try to solve performance issues.
```

=== To capture the logs

. Capture a snapshot of the 'Mobile Developer Console Application' Grafana dashboard and track it over time. The metrics can be useful for identifying performance issues over time.
. Capture the Service logs
.. Get the pod names by running `oc get pods`
.. Save the logs by running `oc logs <mdc-podname> > <filename>.log` for any pod
. Capture CR Status by running `oc describe MobileDeveloperConsole`.

=== To scale the pod

. Scale the Mobile Develper Console pod

> .. Get the replication controller name by running `oc get rc | grep mdc`

Reviewer 2: ```suggestion
```
- See my suggestion on line 79, you need neither the RC name nor the DC name, since we can get it by label
- The RC is kind of an implementation detail that we usually don't need to deal with

> .. Check the number of DESIRED and CURRENT replicas

Reviewer 2: ```suggestion
.. Check the number of DESIRED and CURRENT replicas, and make sure that they're the same:

\```
oc get deploymentconfig --selector='app=mdc'
NAME      REVISION   DESIRED   CURRENT   TRIGGERED BY
mdc       3          1         1         config,image(mdc-imagestream:1.1.4),image(mdc-oauth-proxy-imagestream:latest)
\```

If they're not, this may indicate a problem with the ability to scale up (capacity, missing images, etc), or the ability to scale down (a node where a pod was scheduled might be in trouble or something).
```

^ if accepting this suggestion, remove the escapes for the code block delimiters -- I had to put them in since the `suggestion` feature also uses those.

> .. Scale the replication controller if more pods are desired by running `oc scale rc example-mdc-1 --replicas=x` where x is the number of desired pods.

Reviewer 2: ```suggestion
.. Scale the replication controller if more pods are desired by running `oc scale deploymentconfig --selector='app=mdc' --replicas=X` where x is the number of desired pods.
```
